File: eclipseplugin/com.microsoft.javapkgsrv/src/com/microsoft/javapkgsrv/PipeChannel.java
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT License.  See LICENSE file in the project root for license information.

package com.microsoft.javapkgsrv;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.channels.Channels;

import com.google.protobuf.CodedInputStream;
import com.google.protobuf.CodedOutputStream;

public class PipeChannel {
	public String PipeName = "javapkgsrv";
	private CodedOutputStream CPipeOut = null;
	private CodedInputStream CPipeIn = null;
	private RandomAccessFile Pipe = null;
	
	public PipeChannel()
	{
	}
	public PipeChannel(String pipeName)
	{
		PipeName = pipeName;
	}
	public void Init() throws FileNotFoundException
	{
		Pipe = new RandomAccessFile("\\\\.\\pipe\\" + PipeName, "rw");
		CPipeOut = CodedOutputStream.newInstance(Channels.newOutputStream(Pipe.getChannel()));
		CPipeIn = CodedInputStream.newInstance(Channels.newInputStream(Pipe.getChannel()));		
	}
	public Protocol.Request ReadMessage() throws IOException
	{
		int len = CPipeIn.readInt32();
		byte[] msgBytes = CPipeIn.readRawBytes(len);
		Protocol.Request msg = Protocol.Request.parseFrom(msgBytes);
		return msg;		
	}
	public void WriteMessage(Protocol.Response msg) throws IOException
	{
		byte[] msgData = msg.toByteArray();
		CPipeOut.writeInt32NoTag(msgData.length);
		CPipeOut.writeRawBytes(msgData);
		CPipeOut.flush();		
	}
	public void Disconnect() throws IOException
	{
		Pipe.close();
	}	
}

File: javapkg/javapkg/PipeChannel.cs
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT License.  See LICENSE file in the project root for license information.

using ProtoBuf;
using System;
using System.Collections.Generic;
using System.IO.Pipes;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace javapkg
{
    class PipeChannel : IDisposable
    {
        public string PipeName { get; private set; }
        private NamedPipeServerStream PipeServer { get; set; }
        public PipeChannel()
        {
            PipeName = "javapkgsrv";
        }
        public PipeChannel(string pipeName)
        {
            PipeName = pipeName;
        }
        public PipeStream Init()
        {
            PipeServer = new NamedPipeServerStream(PipeName, PipeDirection.InOut, 4, PipeTransmissionMode.Byte);
            return PipeServer;
        }
        public void WaitForConnection()
        {
            PipeServer.WaitForConnection();
        }
        public Protocol.Response ReadMessage()
        {
            return Serializer.DeserializeWithLengthPrefix<Protocol.Response>(PipeServer, PrefixStyle.Base128);
        }
        public void WaitForDrain()
        {
        }
        public void WriteMessage(Protocol.Request request)
        {
            Serializer.SerializeWithLengthPrefix(PipeServer, request, PrefixStyle.Base128);
        }
        public void Disconnect()
        {
            if (PipeServer == null)
                return;

            PipeServer.Dispose();
            PipeServer = null;
        }
        private bool disposed = false;
        public void Dispose()
        {
            if (!disposed)
            {
                Disconnect();
                disposed = true;
            }
        }
    }
}
